src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -538,6 +545,38 @@ def compile(
             },
         )()
 
+    def always_do_after_compile(self, ctx):
+        self._pass_diagnostics_to_buildstats(ctx)
+
+    def _aggregate_diagnostics(self, lsp_data):
+        # Note: this is not arbitrary. It is exactly every value that the LSP's DiagnosticSeverity allows.
+        # See https://microsoft.github.io/language-server-protocol/specification#diagnostic
+        counts = {
+            "Error": 0,
+            "Warning": 0,
+            "Information": 0,
+            "Hint": 0,
+        }
+        for published_diagnostics in lsp_data:
+            for diagnostic in published_diagnostics["diagnostics"]:
+                severity = diagnostic["severity"]
+                counts[severity] += 1
+        return counts
+
+    def _pass_diagnostics_to_buildstats(self, ctx):
+        diagnostics_file = self._diagnostics_out(ctx)
+        if not os.path.exists(diagnostics_file):
+            return
+        with open(diagnostics_file) as json_diagnostics:
```

Comment:
I'm slightly concerned about the unconditional cost here on cache hits, but hopefully diagnostics are rare enough that this isn't a problem...

Reply:
I would say, let's monitor it and if we see a performance regression, let's feature flag the passing of the `-diag` option. If you don't pass it, this file doesn't exist so this is basically zero cost (minus checking that the one file exists which I guess should be fast compared to fetching the cached results and what not)

Reply:
I agree with at least putting it behind a pants option.

Reply:
Thanks. Did that (also made it advanced and off by default to avoid clutter. We can enable it on CI where we care about it)